x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/logging/Throttler.java

```diff
@@ -9,150 +9,185 @@
             if (value == null) {
-                return new LogExecutor(clock, consumer);
+                return new LogExecutor(clock, logCallback, message);
             }
 
-            return value.compute(consumer, durationToWait);
+            return value;
         });
 
-        // This executes an internal consumer that wraps the passed in one, it will either log the message passed here
-        // unchanged, do nothing if it is in the throttled period, or log this message + some text saying how many times it was repeated
-        logExecutor.log(message);
+        logExecutor.logFirstMessage();
     }
 
     @Override
     public void close() {
         isRunning.set(false);
-        cancellableTask.get().cancel();
+        if (cancellableTask.get() != null) {
+            cancellableTask.get().cancel();
+        }
         logExecutors.clear();
     }
 
-    private static class LogExecutor {
-        private final long skippedLogCalls;
-        private final Instant timeOfLastLogCall;
+    // default for testing
+    static class LogExecutor {
+        // -1 here because we need to determine if we haven't logged the first time
+        // After the first time we'll set it to 0, then the thread that runs on an interval
+        // needs to know if there are any repeated message, if it sees 0, it knows there are none
+        // and skips emitting the message again.
+        private static final long INITIAL_LOG_COUNTER_VALUE = -1;
+
+        private final AtomicLong skippedLogCalls;
+        private Instant timeOfLastLogCall;
         private final Clock clock;
-        private final Consumer<String> consumer;
+        private final Consumer<String> throttledConsumer;
+        private final String originalMessage;
 
-        LogExecutor(Clock clock, Consumer<String> throttledConsumer) {
-            this(clock, 0, throttledConsumer);
+        LogExecutor(Clock clock, Consumer<String> throttledConsumer, String originalMessage) {
+            this(clock, INITIAL_LOG_COUNTER_VALUE, throttledConsumer, originalMessage);
         }
 
-        LogExecutor(Clock clock, long skippedLogCalls, Consumer<String> consumer) {
-            this.skippedLogCalls = skippedLogCalls;
+        LogExecutor(Clock clock, long skippedLogCalls, Consumer<String> throttledConsumer, String originalMessage) {
+            this.skippedLogCalls = new AtomicLong(skippedLogCalls);
             this.clock = Objects.requireNonNull(clock);
             timeOfLastLogCall = Instant.now(this.clock);
-            this.consumer = Objects.requireNonNull(consumer);
+            this.throttledConsumer = Objects.requireNonNull(throttledConsumer);
+            this.originalMessage = Objects.requireNonNull(originalMessage);
         }
 
-        void log(String message) {
-            this.consumer.accept(message);
+        void logRepeatedMessages() {
+            long numSkippedLogCalls;
+            synchronized (skippedLogCalls) {
```

Comment:
The synchronization to is to handle an edge case where one thread is about to increment the counter from 0 to 1 to indicate that the thread should emit the log. Then the thread runs and gets the value as 0 causing it to not emit it but still remove the entry from the map.

I suppose another option would be to leave the entry in the map if the thread doesn't emit it. We could have an ever growing map though if we got a bunch of unique messages that were only logged once and never repeated.

Reply:
Actually maybe I can do this with an atomic update call 🤔